### ISSUE TITLE
Compilation Error [CANCRO]

### ISSUE THREAD

**Issue by Xephi** [MiCode/Xiaomi_Kernel_OpenSource#14]
Again an another error : 

In file included from drivers/tspdrv/tspdrv.c:54:0:
drivers/tspdrv/ImmVibeSPI.c: In function 'drv2604_write_reg_val':
drivers/tspdrv/ImmVibeSPI.c:681:4: error: implicit declaration of function 'KBUILD_STR' [-Werror=implicit-function-declaration]
drivers/tspdrv/ImmVibeSPI.c:681:4: error: 'tspdrv' undeclared (first use in this function)
drivers/tspdrv/ImmVibeSPI.c:681:4: note: each undeclared identifier is reported only once for each function it appears in
drivers/tspdrv/ImmVibeSPI.c: In function 'drv2604_read_reg':
drivers/tspdrv/ImmVibeSPI.c:732:2: error: 'tspdrv' undeclared (first use in this function)
cc1: some warnings being treated as errors

scripts/Makefile.build:307: recipe for target 'drivers/tspdrv/tspdrv.o' failed
make[2]: **\* [drivers/tspdrv/tspdrv.o] Error 1
scripts/Makefile.build:443: recipe for target 'drivers/tspdrv' failed
make[1]: **\* [drivers/tspdrv] Error 2
Makefile:950: recipe for target 'drivers' failed
make: **\* [drivers] Error 2

**Comment by rooque** [comment]
Compiling with what GCC ?

**Comment by Xephi** [comment]
https://android.googlesource.com/platform/prebuilts/gcc/linux-x86/arm/arm-eabi-4.6

**Comment by rooque** [comment]
Try with 4.7

**Comment by Xephi** [comment]
Same error :/

**Comment by loicpirez** [comment]
Xiaomi; it's great to give a kernel, but you could give instruction on how to build too. And which dependancy are necessary. My next phone will be another brand I think.

**Comment by ghost** [comment]
Try GCC 4.9.3 (http://forum.xda-developers.com/showthread.php?p=36677987#post36677987)
I haven't tried it yet though, have more important work than this mess here

**Comment by Xephi** [comment]
Again, same error at the same point :

In file included from drivers/tspdrv/tspdrv.c:58:0:
drivers/tspdrv/ImmVibeSPI.c: In function 'drv2604_write_reg_val':
drivers/tspdrv/ImmVibeSPI.c:681:4: error: implicit declaration of function 'KBUILD_STR' [-Werror=implicit-function-declaration]
drivers/tspdrv/ImmVibeSPI.c:681:4: error: 'tspdrv' undeclared (first use in this function)
drivers/tspdrv/ImmVibeSPI.c:681:4: note: each undeclared identifier is reported only once for each function it appears in
drivers/tspdrv/ImmVibeSPI.c: In function 'drv2604_read_reg':
drivers/tspdrv/ImmVibeSPI.c:732:2: error: 'tspdrv' undeclared (first use in this function)
cc1: some warnings being treated as errors
scripts/Makefile.build:307: recipe for target 'drivers/tspdrv/tspdrv.o' failed
make[2]: **\* [drivers/tspdrv/tspdrv.o] Error 1
scripts/Makefile.build:443: recipe for target 'drivers/tspdrv' failed
make[1]: **\* [drivers/tspdrv] Error 2
Makefile:950: recipe for target 'drivers' failed
make: **\* [drivers] Error 2

**Comment by ghost** [comment]
Remove "-Werror-implicit-function-declaration" from the Makefile in your kernel source directory and try again

**Comment by ghost** [comment]
Kick -Wall out too while your on it. I never had much left for warnings which are treated as errors.. Unless you want to fix every little code derp :P

**Comment by Xephi** [comment]
Tested without theses flags, but always these errors ( cause tspdrv seems to be always undeclared ^^")

**Comment by ghost** [comment]
Yeah, they gave us broken junk.. Theres a file /include/linux/dynamic_debug.h that is unusable.. I imported a clean one from the OnePlus One, which is also a SD801 phone to avoid any issues.. Compiles beyond those errors. Replace your file with this: https://github.com/stefant234/AndromedaCANCRO-KK/blob/master/include/linux/dynamic_debug.h

**Comment by ghost** [comment]
Don't know how I fixed tspdrv, it just stopped quirking :O

**Comment by ghost** [comment]
New place, same error.. Don't bother with my Import.. Something else is wrong here

**Comment by deeplyarnab** [comment]
there are a heck of errors in the source, fixed a lot of them and finally got it buildable but not bootable
try deleting this lines in the original immvibespi.c

> /\* From Xiaomi start _/
>                       pr_debug("drv2604 write 0x%02x, 0x%02x", data[i], data[i
> /_ From Xiaomi end */

i m using the compiler arm-cortex_a15-linux-gnueabihf-linaro_4.9.1-2014.04 http://forum.xda-developers.com/showthread.php?p=59859361#post59859361

**Comment by deeplyarnab** [comment]
These are all the issues i faced and fixed and got it buildable
http://pastebin.com/F0fTegmq

i have got the zImage file but the problem is i dont know how to flash it and tried many methods but it didnt boot and went straight to fastboot

**Comment by ghost** [comment]
In the beginning of the compilation dtb's are generated. http://forum.xda-developers.com/showthread.php?t=2469510
You need to create the boot.img with the additional dt.img for it to boot

**Comment by ghost** [comment]
Already fixed most errors here apparently too.. Going to try your immvibespi.c fix now

**Comment by deeplyarnab** [comment]
i tried that but gave errors, anyways let it be! i got the z image to boot! YAY!!!

used this anykernel zip http://forum.xda-developers.com/showthread.php?t=2670512

**Comment by ghost** [comment]
Full success here too. Meh, I have a different way to assemble zips and boot.imgs :)

**Comment by Xephi** [comment]
I will try theses changes :)  By the way,  is your kernel stable with CM-12 @stefant234 ?

**Comment by ghost** [comment]
I haven't tried, don't have a Mi4 anymore :/ Developing completely blind.. Would you like to give the build a shot? My usual test subjects are busy until the evening..

**Comment by deeplyarnab** [comment]
i m testing my current kernel on mokee cm12, working fine till now

**Comment by loicpirez** [comment]
Is high vibration intensity fix with this one?

**Comment by deeplyarnab** [comment]
dont know about that one, can u explain
i used a different compression method for my kernel and its working fine except 2 issues
1- wifi is not turning on
2- slow paced animations have got laggy, the sppedy ones work fine

**Comment by loicpirez** [comment]
Well,  for exemple in AOSP rom by Ivan for Mi3/Mi4, the vibration is really too much high. When you keep down volume key when you put in Silent mode, you can hear that your vibration does weird sound. It already killed some Mi3 vibrator of users...

**Comment by deeplyarnab** [comment]
the last issue i think that remains is the wifi driver, also videos dont play on the kernel

**Comment by AungThiha** [comment]
@arnabbiswasalsodeep video works fine. only having wifi issue.

**Comment by deeplyarnab** [comment]
then i hv to check cuz in my build the videos dont play, at first they start, then last for 1/4th of a sec and gives error cant play the video
i will do a facorty reset and try

**Comment by emceethemouth** [comment]
I got everything booting with GCC 4.9.3, now all we need is the wifi drivers.

**Comment by schspa** [comment]
Modify the Makefile of  drivers/tspdrv 
the line below will fix the build 
##EXTRA_CFLAGS += -I$(M)
EXTRA_CFLAGS += -I$(shell pwd)
